### PR TITLE
CU-1vz89jk - Liquidity bootstrapping pallet

### PR DESCRIPTION
**Pull request by hussein-aitlahcen**
- based on [balancer](https://balancer.fi/whitepaper.pdf)
- pool can only be created by admin
- only pool owner is able to add liquidity before the sale start
- once the sale started, the weight of the base asset decrease, the initial/final weight can be defined when creating the pool
- once the sale is ended, the owner can withdraw the liquidity, which result in pool deletion

A project token sale example is given in [the plots directory](https://github.com/ComposableFi/composable/tree/4bccda00c460f8a5957623daacb0691fb53ded49/frame/liquidity-bootstrapping/plots)

### CODE REVIEW

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-1vz89jk LBP on top of uniswap](https://app.clickup.com/t/1vz89jk)

File: frame/liquidity-bootstrapping/src/benchmarking.rs [hunk 1]
```diff
@@ -0,0 +1,145 @@
+use super::*;
```
Comment (dzmitry-lahoda) [inline frame/liquidity-bootstrapping/src/benchmarking.rs:1]: may consider including benchamrks as tests as it is done in lending. so they break less and increase coverage.

File: frame/liquidity-bootstrapping/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,725 @@
...
+			let current_quote_weight = Permill::one()
+				.checked_sub(&current_base_weight)
+				.ok_or(ArithmeticError::Underflow)?;
+			Ok((current_base_weight, current_quote_weight))
+		}
+	}
+
+	impl<BlockNumber: Copy + Saturating> Sale<BlockNumber> {
+		pub(crate) fn duration(&self) -> BlockNumber {
+			// NOTE(hussein-aitlahcen): end > start as previously checked by PoolIsValid.
```
Comment (dzmitry-lahoda) [inline frame/liquidity-bootstrapping/src/lib.rs:130]: if it was checked than can use just minus, or just remove comment.

File: frame/liquidity-bootstrapping/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,725 @@
...
+			/*
+			https://balancer.fi/whitepaper.pdf, (15)
+				  ai = amount in = quote_amount
+				  ao = amount out = base_amount
+				  bi = balance in = quote asset balance
+				  bo = balance out = base asset balance
+			wi = weight in
+			wo = weight out
+				   */
+			let (wo, wi) =
```
Comment (dzmitry-lahoda) [inline frame/liquidity-bootstrapping/src/lib.rs:483]: please move into  some math submethod

File: frame/liquidity-bootstrapping/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,725 @@
...
+	}
+
+	impl<BlockNumber: TryInto<u64> + Ord + Copy + Saturating + SafeArithmetic> Sale<BlockNumber> {
+		pub(crate) fn current_weights(
+			&self,
+			current_block: BlockNumber,
+		) -> Result<(Permill, Permill), DispatchError> {
+			/* NOTE(hussein-aitlahcen): currently only linear
+
+			   Assuming final_weight < initial_weight
```
Comment (dzmitry-lahoda) [inline frame/liquidity-bootstrapping/src/lib.rs:101]: please leave comment that base weight diminished with time in docs (so what we sell becomes cheaper in quote)

File: frame/liquidity-bootstrapping/src/lib.rs (hunk 3 above)
Comment (hussein-aitlahcen) [reply frame/liquidity-bootstrapping/src/lib.rs:483]: done!

File: frame/liquidity-bootstrapping/src/benchmarking.rs [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    let project_token: T::AssetId = 0.into();
+    let usdt: T::AssetId = 1.into();
+		let pair = CurrencyPair::new(project_token, usdt);
+		let owner = whitelisted_caller();
+		let pool = Validated::new(Pool {
+			owner,
+			pair,
+			sale: Sale {
+				start: 100,
+				end: 21600 + 100,
```
Comment (andor0) [inline frame/liquidity-bootstrapping/src/benchmarking.rs:23]: maybe add `const SIX_HOURS: u32 = 21600;`?

**Comment by andor0** [comment]
maybe move all PNGs to `frame/liquidity-bootstrapping/docs` directory?

File: frame/liquidity-bootstrapping/src/lib.rs (hunk 4 above)
Comment (hussein-aitlahcen) [reply frame/liquidity-bootstrapping/src/lib.rs:109]: done!

**Comment by hussein-aitlahcen** [comment]
> maybe move all PNGs to `frame/liquidity-bootstrapping/docs` directory?

Yes, probably good to do so!

**Review by dzmitry-lahoda — DISMISSED** [review]

**Comment by hussein-aitlahcen** [comment]
Depends on #618 618

File: frame/liquidity-bootstrapping/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,676 @@
...
+			let pool = Self::get_pool(pool_id)?;
+			ensure!(
+				pool.sale.state(current_block) == expected_sale_state,
+				Error::<T>::InvalidSaleState
+			);
+			Ok(pool)
+		}
+
+		pub(crate) fn get_pool(pool_id: T::PoolId) -> Result<PoolOf<T>, DispatchError> {
+			Pools::<T>::get(pool_id).ok_or(Error::<T>::PoolNotFound.into())
```
[the remark this answers is not in the document]
Comment (hussein-aitlahcen) [reply frame/liquidity-bootstrapping/src/lib.rs:471]: :thinking: did not had any issue with CI?

File: frame/liquidity-bootstrapping/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,735 @@
...
+				pool_id,
+				base_amount,
+				quote_amount,
+				BalanceOf::<T>::zero(),
+				keep_alive,
+			)?;
+			Ok(())
+		}
+
+		///  liquidity to an LBP pool.
```
Comment (vivekvpandya) [inline frame/liquidity-bootstrapping/src/lib.rs:464]: forgot to update comments.

File: frame/liquidity-bootstrapping/src/benchmarking.rs [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+use frame_support::{assert_ok, traits::fungibles::Mutate};
+use frame_system::RawOrigin;
+use sp_arithmetic::Permill;
+
+benchmarks! {
+  where_clause { where T::BlockNumber: From<u32>, T::Balance: From<u128>, T::AssetId: From<u128> }
+
+  create {
+		let unit = 1_000_000_000_000u128;
+	  let project_token: T::AssetId = 0.into();
```
Comment (vivekvpandya) [inline frame/liquidity-bootstrapping/src/benchmarking.rs:15]: minor, indentation should be fixed.

File: frame/liquidity-bootstrapping/src/mock.rs [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+// Configure a mock runtime to test the pallet.
+frame_support::construct_runtime!(
+	pub enum Test where
+		Block = Block,
+		NodeBlock = Block,
+		UncheckedExtrinsic = UncheckedExtrinsic,
+	{
+		System: frame_system::{Pallet, Call, Config, Storage, Event<T>},
+		Tokens: orml_tokens::{Pallet, Call, Storage, Config<T>, Event<T>},
+	CurrencyFactory: pallet_currency_factory::{Pallet, Storage, Event<T>},
```
Comment (vivekvpandya) [inline frame/liquidity-bootstrapping/src/mock.rs:45]: indentation.

File: runtime/dali/Cargo.toml [hunk 10]
```diff
@@ -211,6 +212,7 @@ std = [
 	"simnode-apis/std",
   "uniswap-v2/std",
   "curve-amm/std",
+  "liquidity-bootstrapping/std",
```
Comment (vivekvpandya) [inline runtime/dali/Cargo.toml:215]: indentation for these 3 lines.

File: runtime/dali/Cargo.toml [hunk 11]
```diff
@@ -251,5 +253,6 @@ runtime-benchmarks = [
 	"lending/runtime-benchmarks",
   "uniswap-v2/runtime-benchmarks",
   "curve-amm/runtime-benchmarks",
+  "liquidity-bootstrapping/runtime-benchmarks",
```
Comment (vivekvpandya) [inline runtime/dali/Cargo.toml:256]: same as above

File: runtime/dali/src/weights/mod.rs [hunk 12]
```diff
@@ -17,6 +17,7 @@ pub mod oracle;
 // pub mod scheduler;
 pub mod curve_amm;
 pub mod lending;
+pub mod liquidations;
```
Comment (vivekvpandya) [inline runtime/dali/src/weights/mod.rs:20]: may be forgot to add LBP here?

File: frame/liquidity-bootstrapping/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,735 @@
...
+		///
+		/// The `base_amount` always represent the base asset amount (A/B => A).
+		///
+		/// Emits `Swapped` event when successful.
+		#[pallet::weight(T::WeightInfo::buy())]
+		pub fn buy(
+			origin: OriginFor<T>,
+			pool_id: PoolIdOf<T>,
+			asset_id: AssetIdOf<T>,
+			amount: BalanceOf<T>,
```
Comment (vivekvpandya) [inline frame/liquidity-bootstrapping/src/lib.rs:384]: better to rename `amount` to `base_amount` to be consistent with docs comments.

File: frame/liquidity-bootstrapping/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,735 @@
...
+		///
+		/// The `base_amount` always represent the base asset amount (A/B => A).
+		///
+		/// Emits `Swapped` event when successful.
+		#[pallet::weight(T::WeightInfo::sell())]
+		pub fn sell(
+			origin: OriginFor<T>,
+			pool_id: PoolIdOf<T>,
+			asset_id: AssetIdOf<T>,
+			amount: BalanceOf<T>,
```
Comment (vivekvpandya) [inline frame/liquidity-bootstrapping/src/lib.rs:402]: better to rename `amount` to `base_amount` to be consistent with docs comments.

**Review by vivekvpandya — APPROVED** [review]
Minor: may be added x/y axis description in plots, makes it easier to read.
lbp_buy_project plot has RED and BLUE line overlapping each other, I don't know exact reason but may be because it is small amount of USDT?

File: frame/liquidity-bootstrapping/src/tests.rs [hunk 15]
```diff
@@ -0,0 +1,592 @@
...
+				plot_swap(
+					pair,
+					buy_amount * unit,
+					"./plots/lbp_buy_project.png",
+					format!("Buy project tokens with {} USDT", buy_amount),
+				);
+				let sell_amount = 100_000;
+				plot_swap(
+					pair.swap(),
+					100_000 * unit,
```
Comment (vivekvpandya) [inline frame/liquidity-bootstrapping/src/tests.rs:536]: use `sell_amount` instead of 100_000

File: frame/liquidity-bootstrapping/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,735 @@
...
+
+	#[pallet::storage]
+	#[pallet::getter(fn pools)]
+	pub type Pools<T: Config> = StorageMap<_, Blake2_128Concat, T::PoolId, PoolOf<T>>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Create a new pool.
+		///
+		/// Emits `PoolCreated` even when successful.
```
Comment (vivekvpandya) [inline frame/liquidity-bootstrapping/src/lib.rs:363]: `even` -> `event`

File: frame/liquidity-bootstrapping/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,735 @@
...
+			keep_alive: bool,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let _ = <Self as CurveAmm>::sell(&who, pool_id, asset_id, amount, keep_alive)?;
+			Ok(())
+		}
+
+		/// Execute a specific swap operation.
+		///
+		/// Buy operation if the pair is the original pool pair (A/B).
```
Comment (vivekvpandya) [inline frame/liquidity-bootstrapping/src/lib.rs:412]: I think it is always `sell`.

File: frame/liquidations/src/lib.rs [hunk 18]
```diff
@@ -119,11 +121,11 @@ pub mod pallet {
 	#[pallet::call]
 	impl<T: Config> Pallet<T> {
 		#[pallet::weight(T::WeightInfo::add_liquidation_strategy())]
-		pub fn add_liqudation_strategy(
+		pub fn add_liquidation_strategy(
 			_origin: OriginFor<T>,
 			_configuraiton: LiquidationStrategyConfiguration<T::ParachainId>,
 		) -> DispatchResultWithPostInfo {
-			Err(DispatchError::Other("add_liqudation_strategy: no implemented").into())
+			Err(DispatchError::Other("add_liquidation_strategy: no implemented").into())
```
Comment (KaiserKarel) [inline frame/liquidations/src/lib.rs:128]:
```suggestion
			Err(DispatchError::Other("add_liquidation_strategy: not implemented").into())
```

**Review by KaiserKarel — APPROVED** [review]

**Review by vivekvpandya — APPROVED** [review]

**Pull request merged**